### ISSUE TITLE
multipart: NextPart: EOF error for multipart/mixed content type

### ISSUE THREAD

**Issue by swagftw** [emersion/go-message#174]
I was trying the read the mail using go-imap and go-message libraries, but for a particular mail I am getting
`multipart: NextPart: EOF` error on `part, err := reader.NextPart()` where reader is the `mail.CreateReader`.

Content-Type: multipart/mixed; boundary="-----SECBOUND"

```
-------SECBOUND
Content-Type: text/html
Content-Transfer-Encoding: 8bit

html content only

-------SECBOUND--
```

**Comment by emersion** [comment]
What is the code reading this message?

`NextPart` will return EOF when there are not more parts to read.

**Comment by swagftw** [comment]
```go
part, err := reader.NextPart()
if err == io.EOF {
	break
} else if err != nil && !message.IsUnknownCharset(err) {
	break
}
```

From so long I was thinking that this is a bug because I was already handling the io.EOF error, but as you said it should be just another `EOF` error, I dug deeper. 

The code is pretty straight forward and I am not doing anything fancy here. The error is getting populated through https://github.com/emersion/go-message/blob/8138ff9fec34e40fc969e98c7e6f5b76baff30d7/textproto/multipart.go#L262

But I am not sure why it is not just another `io.EOF` even though the whole body itself is of multipart/mixed type. Also we don't have a concrete error type to check/match the error if this occurs.

**Comment by emersion** [comment]
I've tried running your code with your source message, but I don't get the error: https://go.dev/play/p/cSU0T0XgRL7

**Comment by rexxDigital** [comment]
This still happens by the way, got me confused. when just returning an io.EOF would be sufficient.

**Comment by emersion** [comment]
Please post a playground link with a reproducer, if you can reproduce.

**Comment by rexxDigital** [comment]
> Please post a playground link with a reproducer, if you can reproduce.

Cant post with the actual mail since its private info. But this happens rarely with mails received from certain platforms. Wouldnt returning an io.EOF be just fine here? or create an error type like MultipartEOF?

**Comment by emersion** [comment]
No, the error happens when EOF is unexpected in the middle of a multipart message, so it doesn't make sense to ignore it or treat it like a regular EOF.

**Comment by rexxDigital** [comment]
> No, the error happens when EOF is unexpected in the middle of a multipart message, so it doesn't make sense to ignore it or treat it like a regular EOF.

I see, from my own testing when this happens we have already gotten the body in both html and plain text. Letting the loop run ends in it never breaking as it does not reach the io.EOF return. Having an error type for this would good so we can atleast check for it.

**Comment by emersion** [comment]
If you don't care about errors after the parts you're interested in, you can stop the loop as soon as you've seen these parts.
